Migration to Material Design 3 (fixes #1833, fixes #1418)

Bnyro opened the pull request:
The PR contains the following:
* Monochrome icon (closes #1833)
* Migration to Material Design 3, including dynamic colors if available

also closes #1418 

It might be possible that I forgot to migrate some more hidden parts of the app, feel free to let me know if that's the case so that it can be migrated too!

imsodin commented:
Thanks a lot for your contribution!

In the spirit of transparancy disclaimer up-front: This app is basically on life support. I am even struggling to find the time/energy to keep it building with new android/syncthing/go versions (currently build with go 1.20 is broken). On a very high level you changes look clean, and luckily we have a few motivated test users - so if everything looks and behaves fine we can move forward but I likely won't be able to give much assistance with any problems. Though it looks like you are quite proficient with android apps, so I hope you'll be just fine there.

CI is failing due to this - you can check yourself, you don't need to create an account there (there's some guest access link on the login page):

> [49 ](https://build.syncthing.net/buildConfiguration/SyncthingAndroid_Build/136716?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=136716_1692_1692&logView=flowAware)
>   /opt/tcagent/syncthing-3-work/19a8ed19ae4d1032/app/src/main/java/com/nutomic/syncthingandroid/activities/QRScannerActivity.java:62: Error: Overriding method should call super.onRequestPermissionsResult [MissingSuperCall]
> 
> [19:11:49 ](https://build.syncthing.net/buildConfiguration/SyncthingAndroid_Build/136716?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=136716_1693_1693&logView=flowAware)
>       public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {

Though I would have expected it failing due to bumping of the compile/target SDK: That both requires some changes to the build setup and I'd expect also some changes to the code to adapt to SDK changes. Is this a requirement for the material migration? If not I'd suggest to just roll that back.

Bnyro commented:
Thanks for the feedback!

> CI is failing due to this - you can check yourself, you don't need to create an account there (there's some guest access link on the login page):
> 
Alright, I've attempted to fix that now!
> Though I would have expected it failing due to bumping of the compile/target SDK: That both requires some changes to the build setup and I'd expect also some changes to the code to adapt to SDK changes. Is this a requirement for the material migration? If not I'd suggest to just roll that back.
>
Unfortunately we need to use at least SDK 32 to be able to use the newer version of the Material Design library (which is used for Material Design 3), so I've now changed it from 33 to 32 with the hope that this will make it easier for us for now.

I've also just noted that the app introduction screens are not yet migrated to Material Design 3 yet, so I'll give that a shot too now :)

Bnyro commented:
Seems like the checks have passed! :tada:

imsodin commented:
The built setup changes don't automatically apply, a new builder docker image needs to be deployed first. Still a good sign it already builds on the old image. Build tools should likely be bumped to match the new target (32) as well. To make testing/deploying with the new setup simple, would you mind posting separate PRs for the changes, starting with the SDK bump only? I could also take care of that, but probably not in the next two weeks. I'll try to get it done before the next release cycle (1.23.4 resp 1.24.0, syncthing RC will be released on second Tuesday of April and android app usually too). 

Also looks like 32 is just a "minor" upgrade, still android 12 with stuff for larger screens. From a quick look into the docs, there shouldn't be any behaviour changes that materially (haha) change behaviour for this app - maybe stuff breaks/improves for large devices, which honestly I don't care about - so that should be good.

The material UI changes: Is this mainly a dependency bump/migration without big changes or does this actually change the UI and maybe even behaviour? If it does, could you provide some sample screenshots and in case of behaviour changes a quick summary of what changes please.

Thanks again for your work here!

Bnyro commented:
Indeed that also brings in some design changes as per the guidelines of [Material Design 3](https://m3.material.io/). However, these are not that significant and there are no changes to the behavior of the app. Below some screenshots built from the latest commit of the PR:

![Screenshot_20230326-161722_Syncthing](https://user-images.githubusercontent.com/82752168/227782123-d564b7f3-a1e7-4aca-86e7-5eb558868e9d.png)
![Screenshot_20230326-161750_Syncthing](https://user-images.githubusercontent.com/82752168/227782125-1825d0ee-4f36-4abd-a2a1-3f9127bba2ee.png)
![Screenshot_20230326-161802_Syncthing](https://user-images.githubusercontent.com/82752168/227782126-3b95487b-5411-452f-b854-219ab97d5a26.png)
![Screenshot_20230326-161816_Syncthing](https://user-images.githubusercontent.com/82752168/227782128-feb9a71f-af71-4fdb-9bd8-9d62fc51eafe.png)
![Screenshot_20230326-161830_Syncthing](https://user-images.githubusercontent.com/82752168/227782129-4e753781-e6ce-494c-8812-2f657ef403ce.png)
![Screenshot_20230326-161839_Syncthing](https://user-images.githubusercontent.com/82752168/227782130-bb8562b6-bf24-4ee6-91f9-1b7f8cdc0461.png)
![Screenshot_20230326-161845_Syncthing](https://user-images.githubusercontent.com/82752168/227782133-98cf0f38-0276-4316-b18d-7c26d9fdd2f0.png)
![Screenshot_20230326-161856_Syncthing](https://user-images.githubusercontent.com/82752168/227782134-a4752ea9-fcee-4f42-ad94-fe9c94073bc7.png)
![Screenshot_20230326-161952_Syncthing](https://user-images.githubusercontent.com/82752168/227782135-abfb273d-fd88-4782-ad02-e6e3bd383c17.png)

Bnyro commented:
> The built setup changes don't automatically apply, a new builder docker image needs to be deployed first. Still a good sign it already builds on the old image. Build tools should likely be bumped to match the new target (32) as well. To make testing/deploying with the new setup simple, would you mind posting separate PRs for the changes, starting with the SDK bump only? Resp. I can take care of that myself, but maybe not in the next two weeks. I'll try to get it done before the next release cycle (1.23.4 resp 1.24.0, syncthing RC will be released on second Tuesday of April and android app usually too).
>
You mean just creating a separate PR that bumps the SDK and build tools to version 32 (and also fixes that error that caused the builds to fail previously)? I could do that, however I don't really see the benefit of it since there's no advantage in getting in work within a different PR?

imsodin commented:
I just tested, merged and pushed the sdk bump to 32 - no fixes were necessary, that's likely related to the other changes in this PR. Also re-ran CI on this PR with the new builder image -> green. You can rebase on latest main and drop the sdk related build changes in this PR.

Bnyro commented:
I'm sorry for being annoying, but is the thing currently blocking here that some builds don't work properly yet? The CI on GitHub seems to pass, but I'm not familiar with how the Syncthing CI integration is working actually. Though no trouble if it's just because it requires some testers before, I don't mind that, I just want to make sure I didn't do anything wrong that stops us from continuing and you'd be waiting for me all the time :)

Bnyro commented:
Yes, of course, I'm waiting patiently for some questions I can answer and suggestions I can work on :)

imsodin reviewed (commented):
Looks good. Not going to pretend I understand the purpose of all changes, but changes look sensible (and move away form compat/deprecated stuff, so anyway good), screenshot look nice and anything else will come up in testing (resp. hopefully wont).

Added a few comments. For the "why" questions there's no need to go in depth or anything, if there's a simple answer I'd appreciate it, otherwise for android "because that's how it's done" goes a long way.

imsodin commented:
@Bnyro Did you see the review/questions? I am now more or less waiting for you to give me a bit of reassurance an the few questions above and then I'll ship this. And the next RC just dropped, so it's either more or less now or another month until that gets release (a month isn't an issue at all of course, plus I am not complaining - especially not considering how long it took me to react).

Bnyro commented:
Oh, I just saw it, I'm really sorry!
I somehow didn't get notified (or just missed it because I had so many other unimportant notifications), otherwise I would already have reacted earlier :/
I'm be a bit busy currently but I think I'll find time for this in the next days, so we can finally get this done :)

Bnyro commented on app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java:
```diff
@@ -99,6 +104,12 @@ public class MainActivity extends StateDialogActivity
 
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState, @Nullable PersistableBundle persistentState) {
+        DynamicColors.applyToActivityIfAvailable(this);
```
You're right, it's accidentally been added twice which is not needed.

Bnyro commented on app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:
```diff
@@ -257,12 +258,12 @@ private void addBottomDots() {
-        for (int i = 0; i < mDots.length; i++) {
-            mDots[i].setTextColor(colorsInactive[currentPage]);
+        int colorInactive = MaterialColors.getColor(this, R.attr.colorPrimary, Color.BLUE);
+        int colorActive = MaterialColors.getColor(this, R.attr.colorAccent, Color.BLUE);
```
`colorAccent` got deprecated with Material Design 3. Therefore we should probably use `R.attr.colorSecondary` or `R.attr.colorPrimaryVariant` here I guess, I'll see which fits best.

Bnyro commented on app/src/main/res/values/themes.xml:
```diff
@@ -37,26 +46,37 @@
+    <!-- Preference dialog theme -->
+    <style name="ThemeOverlay.App.MaterialAlertDialog.Monet" parent="ThemeOverlay.Material3.MaterialAlertDialog">
+        <item name="alertDialogStyle">@style/MaterialAlertDialog.App</item>
+        <item name="dialogCornerRadius">28dp</item>
```
Yes, we should do that!

Bnyro commented on app/src/main/res/values/themes.xml:
```diff
@@ -22,10 +21,20 @@
+
+        <item name="android:alertDialogTheme">@style/ThemeOverlay.App.MaterialAlertDialog.Monet</item>
+        <item name="alertDialogTheme">@style/ThemeOverlay.App.MaterialAlertDialog.Monet</item>
+        <item name="android:dialogCornerRadius" tools:targetApi="p">28dp</item>
+        <item name="dialogCornerRadius">28dp</item>
```
Yes, that's how I understand it too!

Bnyro commented:
The latest commit should have addressed all your comments.
I've now also added a top bar to the settings activity, because it would probably feel inconsistent to not have one in the settings and the layout would feel very empty without (in my opinion).
